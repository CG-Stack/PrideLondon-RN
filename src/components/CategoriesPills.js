// @flow
import React from "react";
import { View, StyleSheet, ScrollView, PixelRatio } from "react-native";
import LinearGradient from "react-native-linear-gradient";
import type { ViewStyleProp } from "react-native/Libraries/StyleSheet/StyleSheet";
import CategoryPill from "./CategoryPill";
import Text from "./Text";
import text from "../constants/text";

import {
  darkBlueGreyTwoColor,
  eucalyptusGreenColor,
  blackZeroColor,
  blackThirtyColor
} from "../constants/colors";
import type { EventCategoryName } from "../data/event";

type Props = {
  selectedCategories: Set<EventCategoryName>,
  style?: ViewStyleProp,
  onPress?: Function
};

const createAccessibilityLabel = (
  listPrefix: string,
  pillsList: Set<EventCategoryName>
) => {
  return pillsList.size > 0
    ? `${listPrefix} ${[...pillsList].join(", ")}`
    : text.categoryFilterEmpty;
};

class CategoriesPills extends React.PureComponent<Props> {
  static defaultProps = {
    style: {}
  };

  onTouchStart = () => {
    this.shouldTriggerPress = true;
  };

  onScroll = () => {
    this.shouldTriggerPress = false;
  };

  onTouchEnd = (e: any) => {
    if (this.shouldTriggerPress && this.props.onPress) {
      this.props.onPress(e);
    }
  };

  scrollView: ?Object;
  shouldTriggerPress: boolean = false;

  scrollToLastPill = () => {
    if (!this.scrollView) {
      return;
    }

    this.scrollView.scrollToEnd({ animated: false });
  };

  render() {
    const { style, selectedCategories } = this.props;

    return (
      <View
        style={[styles.selectedCategoriesPills, style]}
        accessible
        accessibilityLabel={createAccessibilityLabel(
          text.categoryFilterContents,
          selectedCategories
        )}
      >
        {selectedCategories.size === 0 ? (
          <Text type="h3" style={styles.zeroSelected}>
            {text.zeroSelected}
          </Text>
        ) : (
          <View style={styles.scrollView}>
            <LinearGradient
              style={[styles.scrollShadow, styles.scrollShadowLeft]}
              end={{ x: 0, y: 0.5 }}
              start={{ x: 1.0, y: 0.5 }}
              colors={[blackZeroColor, blackThirtyColor]}
            />
            <ScrollView
              ref={(ref: ?Object) => {
                this.scrollView = ref;
              }}
              horizontal
              showsHorizontalScrollIndicator={false}
              onContentSizeChange={this.scrollToLastPill}
              onTouchStart={this.onTouchStart}
              onScroll={this.onScroll}
              onTouchEnd={this.onTouchEnd}
            >
              {Array.from(selectedCategories).map(name => (
                <CategoryPill
                  key={name}
                  name={name}
                  style={styles.categoryPill}
                  labelForAccessibility={createAccessibilityLabel(
                    text.categoryFilterContents,
                    selectedCategories
                  )}
                />
              ))}
            </ScrollView>
            <LinearGradient
              style={[styles.scrollShadow, styles.scrollShadowRight]}
              end={{ x: 1.0, y: 0.5 }}
              start={{ x: 0, y: 0.5 }}
              colors={[blackZeroColor, blackThirtyColor]}
            />
          </View>
        )}
      </View>
    );
  }
}

const styles = StyleSheet.create({
  selectedCategoriesPills: {
    backgroundColor: darkBlueGreyTwoColor,
    flex: 1,
    alignItems: "center",
    height: 44 * PixelRatio.getFontScale(),
    borderRadius: 4,
    flexDirection: "row"
  },
  categoryPill: {
    flex: 1,
    justifyContent: "center",
    alignItems: "center",
    marginTop: 8,
    marginBottom: 8,
    marginLeft: 6,
    marginRight: 2
  },
  zeroSelected: {
    color: eucalyptusGreenColor,
    paddingTop: 4,
    paddingLeft: 8
  },
  scrollView: {
    width: "100%"
  },
  scrollShadow: {
    position: "absolute",
    top: 0,
    width: 15,
    height: 44 * PixelRatio.getFontScale(),
    zIndex: 10,
    borderRadius: 4
  },
  scrollShadowLeft: {
    left: 0
  },
  scrollShadowRight: {
    right: 0
  }
});

export default CategoriesPills;
