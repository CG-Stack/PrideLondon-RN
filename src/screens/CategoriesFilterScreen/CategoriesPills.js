// @flow
import React from "react";
import { View, StyleSheet, ScrollView } from "react-native";
import Text from "../../components/Text";
import text from "../../constants/text";
import eventCategories from "../../constants/event-categories";
import {
  whiteColor,
  coralColor,
  darkBlueGreyTwoColor,
  eucalyptusGreenColor
} from "../../constants/colors";

export const locale = "en-GB";

type Props = {
  selectedCategories: Set<string>
};

const categoryLabels = (selectedCategories: Set<string>) =>
  Array.from(selectedCategories);

const CategoriesPills = ({ selectedCategories }: Props) => (
  <View style={styles.selectedCategoriesPills}>
    {selectedCategories.size === 0 ? (
      <Text type="h3" style={styles.zeroSelected}>
        {text.zeroSelected}
      </Text>
    ) : (
      <ScrollView horizontal showsHorizontalScrollIndicator={false}>
        {categoryLabels(selectedCategories).map(name => (
          <Text
            key={name}
            type="h3"
            style={[
              styles.categoryPill,
              { backgroundColor: eventCategories[locale][name].color }
            ]}
          >
            {name}
          </Text>
        ))}
      </ScrollView>
    )}
  </View>
);

const styles = StyleSheet.create({
  selectedCategoriesPills: {
    backgroundColor: darkBlueGreyTwoColor,
    padding: 10,
    paddingLeft: 16,
    borderRadius: 4,
    marginBottom: 16,
    display: "flex",
    flexDirection: "row"
  },
  zeroSelected: {
    color: eucalyptusGreenColor,
    paddingTop: 2
  },
  categoryPill: {
    color: whiteColor,
    backgroundColor: coralColor,
    paddingTop: 2,
    paddingLeft: 5,
    paddingRight: 5,
    marginRight: 11
  }
});

export default CategoriesPills;
