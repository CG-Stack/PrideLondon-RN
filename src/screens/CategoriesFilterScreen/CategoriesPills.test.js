// @flow
import React from "react";
import { shallow } from "enzyme";
import CategoriesPills from "./CategoriesPills";

const render = props => shallow(<CategoriesPills {...props} />);

describe("CategoriesPills Component", () => {
  it("renders correctly", () => {
    expect(render({ selectedCategories: new Set() })).toMatchSnapshot();
  });

  it("renders no pills", () => {
    const output = render({ selectedCategories: new Set() });
    expect(output.find("Text").props().children).toEqual("0 selected");
  });

  it("renders categories pills", () => {
    const selectedCategories: Set<string> = new Set(["Music", "Community"]);
    const output = render({ selectedCategories });

    expect(output.find("ScrollView").exists()).toEqual(true);
    expect(output.find("ScrollView").children().length).toEqual(
      selectedCategories.size
    );
  });
});
