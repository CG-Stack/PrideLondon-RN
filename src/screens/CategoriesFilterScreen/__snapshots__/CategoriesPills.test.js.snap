// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CategoriesPills Component renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#20215a",
      "borderRadius": 4,
      "display": "flex",
      "flexDirection": "row",
      "marginBottom": 16,
      "padding": 10,
      "paddingLeft": 16,
    }
  }
>
  <Text
    markdown={false}
    style={
      Object {
        "color": "#2cda9d",
        "paddingTop": 2,
      }
    }
    type="h3"
  >
    0 selected
  </Text>
</View>
`;
